### PR TITLE
rock 1.0.22 - batch one

### PR DESCRIPTION
**Pull request by thomasfanell**
**NOTE**: Removed bounty [extend generic classes](https://github.com/magic-lang/rock/pull/44) for now, since it breaks compatibility. It can be found (isolated) in [this branch](https://github.com/thomasfanell/rock/tree/rock_1.0.22_extend_generics).

**Changes**
- C generator outputs `_this` instead of `this`
- ~~Updated obfuscator~~ next batch
- Use of `super()` in a shadowed function now cause an error (use `--allow-super-when-shadowing` to disable this error)
- Auto-generate `count()` and `values()` functions for enums

### CODE REVIEW

File: source/rock/RockVersion.ooc [hunk 1]
```diff
@@ -4,9 +4,9 @@ RockVersion: class {
 
     getMajor:    static func -> Int    { 1 }
     getMinor:    static func -> Int    { 0 }
-    getPatch:    static func -> Int    { 21 }
+    getPatch:    static func -> Int    { 22 }
     getRevision: static func -> String { "head" }
-    getCodename: static func -> String { "Det. Freppowicz" }
+    getCodename: static func -> String { "<not yet>" }
```
Comment (marcusnaslund) [inline source/rock/RockVersion.ooc]: This will be exciting.

Comment (thomasfanell) [inline source/rock/RockVersion.ooc]: Still haven't decided if I want to keep the captain as my primary target, or maybe shift focus to some code reviewer...

Comment (marcusnaslund) [inline source/rock/RockVersion.ooc]: :scream:

Comment (marcusnaslund) [inline source/rock/RockVersion.ooc]: I'm honoured.

Comment (vendethiel) [inline source/rock/RockVersion.ooc]: Marcus Freppowicz... wait.

**Comment by vendethiel** [comment]
> since it breaks compatibility.

FMI, what's broken with that?

**Comment by sebastianbaginski** [comment]
I like how you removed 300 lines from the middle end and it still works :+1:

**Comment by thomasfanell** [comment]
@vendethiel 
All the class extensions have to be changed from `extend <Conrete Type>` to `extend <T>`. rock can not handle both.

**Comment by vendethiel** [comment]
So you're removing specialization from OOC, is that right?

**Comment by thomasfanell** [comment]
@vendethiel no, not really. You can still extend classes, but only by specifying the concrete type, just like in the official ooc-lang, before our bounty-program.

**Comment by vendethiel** [comment]
Okay, it's the opposite : you need to specialize a generic to add behavior.

What was the compatibility issue?

**Comment by thomasfanell** [comment]
@vendethiel 

In rock 1.0.21 (the current version), we extend classes using this syntax: `extend Cell<SomeType>`
The generic extension bounty would allow us to do this instead:  `extend Cell<T>`

Neat, but it requires all of our current extensions to be updated according to the new syntax and we can no longer test our code base on previous versions of rock. We are not ready to commit to this change at this point in time.

**Comment by vendethiel** [comment]
You confirm the fact that `extend Cell<Int>` means these fonctions only exist for `Cell<Int>`, not for any type `T` – like the bounty tries to achieve? I don't see why these should be conflicting features (they're not in, e.g., C++)

**Comment by thomasfanell** [comment]
Yes, that is correct.

For any technical reasons, you'd have to ask @shamanas. There is some discussion in #44.

**Comment by vendethiel** [comment]
Okay. I'd argue this should have been `extend <T> Cell<T>`, but ships sail :)

**Comment by thomasfanell** [comment]
@fredrikbryntesson review this first batch of changes?

**Pull request merged**